tools/istio-iptables/pkg/validation/validator.go

```diff
@@ -0,0 +1,121 @@
+package validation
+
+import (
+	"errors"
+	"fmt"
+	"istio.io/istio/tools/istio-iptables/pkg/config"
+	"istio.io/istio/tools/istio-iptables/pkg/constants"
+	"net"
+	"strconv"
+	"time"
+)
+
+type Validator struct {
+	Config *ValidationConfig
+}
+
+type ValidationConfig struct {
+	ServerListenAddress string
+	ServerOriginalPort  uint16
+	ServerOriginalIp    net.IP
+}
+type ValidationService struct {
+	Config *ValidationConfig
+}
+type ValidationClient struct {
+	Config *ValidationConfig
+}
+
+func (validator *Validator) Run() error {
+	s := ValidationService{
+		validator.Config,
+	}
+	sError := make(chan error, 1)
+	sTimeout := time.After(30 * time.Second)
+	go func() {
+		sError <- s.Run()
+	}()
+
+	// infinite loop
+	go func() {
+		c := ValidationClient{Config: validator.Config}
+		for {
+			c.Run()
+			time.Sleep(time.Second)
+		}
+	}()
+	select {
+	case <-sTimeout:
+		return errors.New("validation timeout")
+	case err := <-sError:
+		if err == nil {
+			fmt.Println("validation passed!")
+		} else {
+			fmt.Println("validation failed:" + err.Error())
+		}
+		return err
+	}
+}
+
+func NewValidator(config *config.Config) *Validator {
+	return &Validator{
+		Config: &ValidationConfig{
+			ServerListenAddress: ":" + config.InboundCapturePort,
+			ServerOriginalPort:  constants.IPTABLES_PROBE_PORT,
+			ServerOriginalIp:    config.HostIp,
+		},
+	}
+}
+
+// Write human readable content
+func echo(conn net.Conn, echo []byte) {
+	conn.Write(echo)
+	conn.Close()
+}
+
+func (s *ValidationService) Run() error {
+	l, err := net.Listen("tcp", s.Config.ServerListenAddress)
+	if err != nil {
+		fmt.Println("Error on listening:", err.Error())
+		return err
+	}
+	// Close the listener when the application closes.
+	defer l.Close()
+	fmt.Println("Listening on " + s.Config.ServerListenAddress)
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			fmt.Println("Error accepting: ", err.Error())
+			continue
+		}
+		_, port, err := GetOriginalDestination(conn)
```

Comment:
Don’t know why we need to do this.
We need to ensure that we send the request out to a port that we are sure must be captured based on parameters, then open an http connection to that port.